### PR TITLE
Make import.num_threads be able to config online

### LINKED ISSUE #17572: br: make import.num-threads dynamic
### Is your feature request related to a problem? Please describe:

currently, lightning and br restore share tikv parameters under `import` section.  And import.num-threads change involves tikv reboot.

### Describe the feature you'd like:

we need to make import.num-threads dynamic

### PR DESCRIPTION
**Pull request by RidRisR**
### What is changed and how it works?

Issue Number: Close #17572

What's Changed:

We add a new thread pool in utils named `ResizableRuntime` which is able to resize itself without killing the running tasks.
With this new thread pool, `import.num_threads` is able to change online without restart the tikv server.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

```
[root@tikv]# ps -TC tikv-server -o comm | uniq -c
      #...
      8 import
      #...
[root@tikv]# tikv-ctl --host 127.0.0.1:20160 modify-tikv-config -n import.num_threads -v 4
success
[root@tikv]# ps -TC tikv-server -o comm | uniq -c
      #...
      4 import
      #...
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Make `import.num_threads` be able to change online without restart the tikv-server.
```

### CODE REVIEW

File: components/tikv_util/src/resizable_threadpool.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        // it is safe because all tasks should be finished.
+        self.0.take().unwrap().shutdown_background()
+    }
+}
+
+pub struct ResizableRuntime {
+    pub size: usize,
+    thread_name: String,
+    pool: Option<Arc<DaemonRuntime>>,
+    after_adjust: fn(usize),
```
Comment (YuJuncen) [inline components/tikv_util/src/resizable_threadpool.rs:40]: Use `Box<Fn(usize)>` instead. `fn(usize)` is a function pointer. Just consider the difference between `std::function<void(int)>` and `void (*fp)(int)`.

File: components/tikv_util/src/resizable_threadpool.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            thread_name: thread_name.to_owned(),
+            pool: None,
+            after_adjust,
+            replace_pool_rule,
+        }
+    }
+
+    pub fn spawn<Func>(&self, func: Func)
+    where
+        Func: Future<Output = ()> + Send + 'static,
```
Comment (YuJuncen) [inline components/tikv_util/src/resizable_threadpool.rs:61]: Why this is a `Future` but its type named `Func`?

File: components/tikv_util/src/resizable_threadpool.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    pub fn block_on<F>(&self, f: F) -> F::Output
+    where
+        F: Future,
+    {
+        let inner = self.inner.read().unwrap();
+        inner.block_on(f)
+    }
+}
+
+pub trait TokioRuntimeCreator {
```
Comment (YuJuncen) [inline components/tikv_util/src/resizable_threadpool.rs:131]: It seems nowhere requires this trait. Consider remove it.

File: components/sst_importer/src/sst_importer.rs [hunk 4]
```diff
@@ -2288,8 +2293,23 @@ mod tests {
         };
         let change = cfg.diff(&cfg_new);
 
+        struct TestImportRuntimeCreator;
+        impl TokioRuntimeCreator for TestImportRuntimeCreator {
+            fn create_tokio_runtime(_: usize, _: &str) -> TokioResult<Runtime> {
```
Comment (YuJuncen) [inline components/sst_importer/src/sst_importer.rs:2298]: `create_tokio_runtime` exists everywhere. Please extract it to a common space.

File: components/sst_importer/src/import_mode.rs [hunk 5]
```diff
@@ -120,6 +120,39 @@ impl ImportModeSwitcher {
         executor.spawn(timer_loop);
     }
 
+    // start_resizable_threads only serves for resizable runtime
+    pub fn start_resizable_threads<E: KvEngine>(&self, executor: &ResizableRuntimeHandle, db: E) {
```
Comment (YuJuncen) [inline components/sst_importer/src/import_mode.rs:124]: Either remove the old `start` function or reuse it.

File: src/import/sst_service.rs [hunk 6]
```diff
@@ -321,47 +319,39 @@ impl<E: Engine> ImportSstService<E> {
...
-                },
-                move || {
-                    // SAFETY: we have set the engine at some lines above with type `E`.
-                    unsafe { tikv_kv::destroy_tls_engine::<E>() };
-                },
-            )
-            .build()
-            .unwrap();
+        // let props = tikv_util::thread_group::current_properties();
+        // let eng = Mutex::new(engine.clone());
```
Comment (YuJuncen) [inline src/import/sst_service.rs:323]: Please remove the temporary lines.

**Review by YuJuncen — APPROVED** [review]
rest lgtm

**Comment by RidRisR** [comment]
@ zhangjinpeng87

**Comment by RidRisR** [comment]
@ zhangjinpeng87

**Review by Leavrth — APPROVED** [review]

**Review by zhangjinpeng87 — APPROVED** [review]
LGTM for components/sst_importer/src/config.rs

**Pull request merged**